java/libs/src/main/java/org/physical_web/collection/UrlDevice.java

```diff
@@ -159,49 +159,153 @@ public String getUrl() {
+   *     coerced into the necessary type.
+   * @return The stored value or null if it doesn't exist in specified form.
+   */
+  public int optExtraInt(String key, int defaultValue) {
+    return mExtraData.optInt(key, defaultValue);
+  }
+
+  /**
    * Get extra long value.
+   * @param key The key of the stored value.
    * @return The stored value or null if it doesn't exist in specified form.
    */
-  public double getExtraLong(String key) {
+  public long getExtraLong(String key) {
+    return mExtraData.getLong(key);
+  }
+
+  /**
+   * Get extra long value.
+   * @param key The key of the stored value.
+   * @return The stored value or null if it doesn't exist in specified form.
+   */
+  public long optExtraLong(String key) {
     return mExtraData.optLong(key);
   }
 
   /**
+   * Get extra long value.
+   * @param key The key of the stored value.
+   * @param defaultValue The value to return if the key does not exist or the value cannot be
+   *     coerced into the necessary type.
+   * @return The stored value or null if it doesn't exist in specified form.
+   */
+  public long optExtraLong(String key, long defaultValue) {
+    return mExtraData.optLong(key, defaultValue);
+  }
+
+  /**
    * Get extra double value.
+   * @param key The key of the stored value.
    * @return The stored value or null if it doesn't exist in specified form.
    */
   public double getExtraDouble(String key) {
+    return mExtraData.getDouble(key);
+  }
+
+  /**
+   * Get extra double value.
+   * @param key The key of the stored value.
+   * @return The stored value or null if it doesn't exist in specified form.
+   */
+  public double optExtraDouble(String key) {
     return mExtraData.optDouble(key);
   }
 
   /**
+   * Get extra double value.
+   * @param key The key of the stored value.
+   * @param defaultValue The value to return if the key does not exist or the value cannot be
+   *     coerced into the necessary type.
+   * @return The stored value or null if it doesn't exist in specified form.
+   */
+  public double optExtraDouble(String key, double defaultValue) {
+    return mExtraData.optDouble(key, defaultValue);
+  }
+
+  /**
    * Get extra JSONArray value.
+   * @param key The key of the stored value.
    * @return The stored value or null if it doesn't exist in specified form.
    */
   public JSONArray getExtraJSONArray(String key) {
+    return mExtraData.getJSONArray(key);
+  }
+
+  /**
+   * Get extra JSONArray value.
+   * @param key The key of the stored value.
+   * @return The stored value or null if it doesn't exist in specified form.
+   */
+  public JSONArray optExtraJSONArray(String key) {
     return mExtraData.optJSONArray(key);
   }
 
   /**
    * Get extra JSONObject value.
-   * @return The stored value or null if it doesn't exist in specified form.
+   * @param key The key of the stored value.
+   * @return The stored value.
    */
   public JSONObject getExtraJSONObject(String key) {
```

Comment:
addExtra(String, Object) can add "Any object of type JSONObject, JSONArray, String, Boolean, Integer, Long, Double, NULL, or null. May not be NaNs or infinities" but currently we have no way to retrieve a string value.